### ISSUE TITLE
`dart pub unpack` fails with dependency published from a workspace

### ISSUE THREAD

**Issue by Luckey-Elijah** [dart-lang/pub#4393]
# Environment
```
Dart SDK version: 3.5.3 (stable) (Wed Sep 11 16:22:47 2024 +0000) on "linux_x64"
```

# Problem
 
`unpack` command runs `dart pub get` in dependency that is a workspace and fails.

## Expected behavior

`unpack` should work "out of the box" with `resolution: workspace` packages

`dart pub unpack binary_serializable` should succeed.

## Actual behavior
```
$ dart pub unpack binary_serializable
Downloading binary_serializable 0.3.0 to `./binary_serializable-0.3.0`... 
To explore type: cd ./binary_serializable-0.3.0
Found a pubspec.yaml at /home/elijah/Development/binary_serializable-0.3.0. But it has resolution `workspace`.
But found no workspace root including it in parent directories.

See https://dart.dev/go/pub-workspaces for more information.
```

The exit code is `66`

I am able to resolve by removing the `resolution: workspace` entry in the `pubspec.yaml` and running `dart pub get` by hand.

**Comment by sigurdm** [comment]
Yeah - this is a bit unfortunate.

This speaks for having a resolution mode that ignores the `resolution: workspace`... or we could explicitly remove the `resolution: workspace` when unpacking.

I don't really like any of those solutions. `unpack` changing the files seem surprising. Resolving dependencies in a non-standard mode also seems surprising.

We could have a command `dart pub workspace eject` that removes the package from the workspace, and suggest that command when the resolution fails.

@jonasfj what do you think...

**Comment by jonasfj** [comment]
Nothing is fully thought through.

--------------

> what do you think...

Off the top my head, it makes me wonder if adding `resolution: workspace` was necessary.
Maybe we could just have searched upwards until we found nothing or a `pubspec.yaml` with a `workspace` property.

Since we ended up doing the `.dart_tool/workspace_ref.json` (or something like that), for caching purposes anyways.

---------------------------

But in a world where we don't undo this, I agree, it's tempting to make a mode that ignores `resolution: workspace`.

But I fear that it would poorly as opening in the IDE often triggers a `pub get` (certainly editing `pubspec.yaml` does), as does `dart run` or `dart test`.

In general: We want `dart pub get` to be safe to run implicitly.

If `dart pub get` is always safe to run, then it's hard to have modes. Unless we persist the `mode` in `pubspec.lock`. But then we're introducing an entirely new concept. And it's unclear to me that persisting mode in `pubspec.lock` won't cause confusing and weird scenarios where users don't understand the behavior.

-----------------

> I don't really like any of those solutions. `unpack` changing the files seem surprising. Resolving dependencies in a non-standard mode also seems surprising.

Agreed, but doing so in order to do `dart pub get` might be okay, if it prints a warning line.

It's certainly an option to consider.

Could we do it by adding a `pubspec_override.yaml` ? (are `pubspec_override.yaml` files ever published?)
